### PR TITLE
core(lantern): fallback to FCP in 0-weight SI situations

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
I was finally able to reproduce https://github.com/GoogleChrome/lighthouse/issues/11154 and found the issue. In some circumstances in DevTools only under desktop throttling (i.e. no CPU throttling), we end up with no consequential layout events (this trace has a single layout event that lasted 1ms). Because we use the log magnitude, log 1 is 0 which can't be a denominator :)

**Related Issues/PRs**
fixes https://github.com/GoogleChrome/lighthouse/issues/11154

### CODE REVIEW

File: lighthouse-core/computed/metrics/lantern-speed-index.js [hunk 1]
```diff
@@ -134,14 +134,12 @@ class LanternSpeedIndex extends LanternMetric {
...
-    if (!layoutWeights.length) {
-      return fcpTimeInMs;
-    }
-
     const totalWeightedTime = layoutWeights
       .map(evt => evt.weight * Math.max(evt.time, fcpTimeInMs))
       .reduce((a, b) => a + b, 0);
     const totalWeight = layoutWeights.map(evt => evt.weight).reduce((a, b) => a + b, 0);
+
+    if (!totalWeight) return fcpTimeInMs;
```
Comment (patrickhulce) [inline lighthouse-core/computed/metrics/lantern-speed-index.js:142]: if `layoutWeights.length === 0` then totalWeight will still be `0`

File: lighthouse-core/test/computed/metrics/speed-index-test.js [hunk 2]
```diff
@@ -23,7 +25,46 @@ describe('Metrics: Speed Index', () => {
       timing: Math.round(result.timing),
       optimistic: Math.round(result.optimisticEstimate.timeInMs),
       pessimistic: Math.round(result.pessimisticEstimate.timeInMs),
-    }).toMatchSnapshot();
+    }).toMatchInlineSnapshot(`
+      Object {
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/metrics/speed-index-test.js:29]: we initially committed this snapshot before jest supported inline snapshots, which are greatly preferred for objects of this size to make the expectations more obvious and easier to notice

Comment (connorjclark) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: aside: did jest mess with the format of the rest of the file? If so, could you try updating to the latest jest and see if they've fixed it? I know it was on their radar.

Comment (patrickhulce) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: prettier's API doesn't support it yet https://github.com/prettier/prettier/issues/5807

it only affected the 1 line that would be updated on any future `-u` so didn't seem like a big deal 🤷‍♂️

Comment (connorjclark) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: darn, still not done.

it's a much larger annoyance for other files, requiring you to either change dozens of other lines to prettier's standards, or copy and pasting the result `-u`, revert everything, and individually applying the new snapshot.

Comment (patrickhulce) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: It is, and I'd obvs love to adopt prettier even if it's only in our test files, but some `git checkout -p` to clean it up definitely beats manually updating all the snapshot expectations :)

Comment (connorjclark) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: nice, didn't know checkout could do patches.

Comment (brendankenny) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:29]: > _looks around to see if @brendankenny is hiding in the tall grass_

🌾🌿 ⚔️ 🌿🌾

File: lighthouse-core/computed/metrics/lantern-speed-index.js [hunk 3]
```diff
@@ -134,14 +134,12 @@ class LanternSpeedIndex extends LanternMetric {
...
     return totalWeightedTime / totalWeight;
```
Comment (brendankenny) [inline lighthouse-core/computed/metrics/lantern-speed-index.js:143]: it's hard to get my brain around possible values here without looking at a bunch of examples, but is it possible to get a weird discontinuity where adding a single layout task could give a result *lower* than `fcpTimeInMs`?

IOW, should this line be `return Math.max(totalWeightedTime / totalWeight, fcpTimeInMs);` or is that implicitly handled?

(this isn't something new, handling the division by zero case is 👍 )

File: lighthouse-core/test/computed/metrics/speed-index-test.js [hunk 4]
```diff
@@ -23,7 +25,46 @@ describe('Metrics: Speed Index', () => {
...
+
+    expect({
+      timing: Math.round(result.timing),
+      optimistic: Math.round(result.optimisticEstimate.timeInMs),
+      pessimistic: Math.round(result.pessimisticEstimate.timeInMs),
+    }).toMatchInlineSnapshot(`
+      Object {
+        "optimistic": 575,
+        "pessimistic": 563,
+        "timing": 599,
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:65]: IMO, not worth it. conceptually speed index *could* be lower than FCP due to background color painting so if we ever decided to try to match that it wouldn't be an assertion that holds up.

File: lighthouse-core/test/computed/metrics/speed-index-test.js [hunk 5]
```diff
@@ -10,6 +10,8 @@ const assert = require('assert').strict;
 const SpeedIndex = require('../../../computed/metrics/speed-index.js');
 const trace = require('../../fixtures/traces/progressive-app-m60.json');
 const devtoolsLog = require('../../fixtures/traces/progressive-app-m60.devtools.log.json');
+const traceNaN = require('../../fixtures/traces/speedindex-nan-m84.trace.json');
+const devtoolsLogNaN = require('../../fixtures/traces/speedindex-nan-m84.devtoolslog.json');
```
[the remark this answers is not in the document]
Comment (patrickhulce) [reply lighthouse-core/test/computed/metrics/speed-index-test.js:14]: `speedindex-1ms-layout` sgtm 👍

File: lighthouse-core/computed/metrics/lantern-speed-index.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/computed/metrics/lantern-speed-index.js:143]: > is it possible to get a weird discontinuity where adding a single layout task could give a result lower than fcpTimeInMs

no the `.map(evt => evt.weight * Math.max(evt.time, fcpTimeInMs))` above should prevent any time earlier than `fcpTimeInMs` from being used in the weight

Comment (brendankenny) [reply lighthouse-core/computed/metrics/lantern-speed-index.js:143]: Oh, yep, there it is :)

**Pull request merged**